### ISSUE TITLE
Test metrics almost zero despite decent training/validation losses.

### ISSUE THREAD

**Issue by naga-karthik** [ivadomed/model_seg_sci#11]
The SoftSeg approach results in near-zero (i.e. `~0`) test evaluation metrics despite showing a decreasing trend in the training and validation loss values. The SoftSeg training approach was taken by observing the GT masks which themselves were between 0 and 1 at the boundaries (as opposed to binary segmentation masks). The necessary parameters were also changed in the config file, particularly, the binarized predictions were disabled and the loss was changed to AdapWingLoss. The list of hyperparameters tuned and the corresponding results obtained are being recorded in [this](https://docs.google.com/spreadsheets/d/1OoGJ5uboMy1JSCIVWGd1k3VAIkmILMHTIMRU_evRPog/edit#gid=0) file. Only Random Reverse augmentation was used in the experiments until now.

As an outlier experiment, instead of training a 3D model, the 2D U-Net was trained on the sagittal slices to observe whether the other planes are somehow hurting the model performance - this can be ruled out as the test metrics for this experiment do not suggest so.

Potential next steps:
1. Train the model with only a few subjects to see if the model overfits. 
2. Check the inputs to the model. One possibility could be that it is not getting the proper inputs, which is causing an issue during testing.

**Comment by uzaymacar** [comment]
I can confirm that I observe the same phenomenon both with `ivadomed` training and custom training. The custom training was performed similar to what is mentioned [here](https://github.com/ivadomed/model_seg_ms_mp2rage/pull/11#issuecomment-992718212). This is important because it confirms that the issue is not related to a potential bug in `ivadomed`.

**Comment by uzaymacar** [comment]
Additionally, I tried training and testing on the same subjects which yielded very good Dice scores -- this shows that indeed the model is **overfitting** and is having a hard time to generalize. Issue #13 as linked above aims to create a discussion on how we can address this.

**Comment by naga-karthik** [comment]
Just some updates based on recent experiments on the full axial-sagittal registered dataset (with 22 subjects): 

1. An issue with multichannel training in ivadomed (documented [here](https://github.com/ivadomed/ivadomed/issues/1087)) was faced - work is being done on this in parallel, however, in the meantime, I am training custom models (code [here](https://github.com/ivadomed/model_seg_sci/tree/nk/single_contrast_training/modeling)) and experimenting with multichannel models.
2. The status of the issue (i.e. test metrics ~0) **still remains**. In particular, there is some heavy overfitting as can be seen in the wandb graphs [here](https://wandb.ai/naga-karthik/sci-zurich/runs/ssq0t6jd?workspace=user-naga-karthik). However, there's also a glimmer of hope - below, I visualized the test predictions on `fsleyes` and I see that the model is getting some parts of the lesions right, and in some cases, it is completely missing the location of the lesion (classic overfitting). Figures:

<details>
<summary> sub-zh10_ses-01 (Prediction in Blue and GT in Red) </summary>

![sub-zh10_ses-01_test](https://user-images.githubusercontent.com/53445351/156220765-45951569-8d59-4cb0-9163-e1912b642e2c.PNG)

</details>

<details>
<summary> sub-zh03_ses-01 (left) and sub-zh22_ses-02 (right) </summary>

![sub-zh03_ses-01_test](https://user-images.githubusercontent.com/53445351/156220974-45d2118d-2042-4dd0-b23a-ae8c08edc928.PNG) | ![sub-zh22_ses-02_test](https://user-images.githubusercontent.com/53445351/156220985-ade6229c-99a4-4cbf-a8d6-32eaf739ed6b.PNG)

</details>

This tells me that the model is at least learning _something_ but not correctly, probably because of the lack of sufficient examples. 

**Next Steps**:

1. Until now, only UNet3D (adapted from `ivadomed.models`) has been used for all the experiments, which I think is deep and also results in over-parametrization. The next step would be to try a simpler model e.g. ResNet3D and see how it performs
2. I was wondering how much it would help to combine the `sci-colorado` with the `sci-zurich` dataset. There are some logistical issues with those which have to be discussed in the next meeting.

**Comment by jcohenadad** [comment]
Thank you for these very interesting updates @naga-karthik !

**Comment by naga-karthik** [comment]
Few updates based on recent experiments with the updated dataset with 73 subjects. Important points to be noted:

1. The dataset is relatively larger with 73 subjects (compared to 22 subjects before). 
2. Preprocessing is also modified in this way - The sagittal T2w 3D images were cropped using the axial T2w 3D image as the reference (because the FoV of the axial scan contains the lesion). The idea was to restrict the model's input to targeted regions containing the lesions.
3. Training is done using `ivadomed` only the cropped sagittal T2w 3D images for the results shown below. 

The issue of ~0 test metrics is longer there, as the average Dice score (across four runs) is 0.246. Detailed results are shown in the table below:

| Seed         | Dice            | IoU             | Hausdorff Score  |
| :--------------: | :-----------------: | :-----------------: | :------------------: |
| 6            | 0.265           | 0.164           | 0.374            |
| 42           | 0.318           | 0.199           | 0.402            |
| 123          | 0.225           | 0.137           | 0.39             |
| 1313         | 0.177           | 0.105           | 0.212            |
| Mean +/- Std | **0.246 +/- 0.052** | **0.151 +/- 0.035** | **0.3445 +/- 0.077** |

Next Steps

1. The method used for cropping ie cropping w.r.t the Axial T2w scan is not generalizable because it excludes the subjects that do not have an axial scan. [Mask R-CNN](https://arxiv.org/pdf/1703.06870.pdf) is an interesting approach that detects the bounding box around the RoI and also segments it. 
    * Work is currently being done to implement this for SCI Lesion segmentation. The hypothesis is that it should be able to better segment the lesion because the RoI is centred around the lesion and importantly, it is more "scalable" as there would be no hard requirement of having axial T2w scans.

**Comment by naga-karthik** [comment]
Closing this as it is very out-of-date (we're using nnUNet and test metrics are not ~0 anymore)
